File: shared/src/main/java/ru/privatenull/pnauth/kernel/event/ListenerOptions.java
package ru.privatenull.pnauth.kernel.event;
public record ListenerOptions(String ownerId, EventPriority priority, boolean receiveCancelled, ListenerMode mode) {
    public ListenerOptions {
        if (ownerId == null || ownerId.isBlank()) throw new IllegalArgumentException("ownerId is required");
        priority = priority == null ? EventPriority.NORMAL : priority;
        mode = mode == null ? ListenerMode.MUTATING : mode;
        if (priority.equals(EventPriority.MONITOR)) mode = ListenerMode.MONITOR;
        if (mode == ListenerMode.MUTATING && priority.value() > EventPriority.MAX_CUSTOM
                && !ownerId.equals("pnauth:system")) throw new IllegalArgumentException("reserved priority");
    }
    public static ListenerOptions normal(String ownerId) {
        return new ListenerOptions(ownerId, EventPriority.NORMAL, false, ListenerMode.MUTATING);
    }
    public static ListenerOptions monitor(String ownerId) {
        return new ListenerOptions(ownerId, EventPriority.MONITOR, true, ListenerMode.MONITOR);
    }
}

File: shared/src/main/kotlin/ru/privatenull/pnauth/kernel/event/ListenerOptions.kt
package ru.privatenull.pnauth.kernel.event

@JvmRecord
data class ListenerOptions @JvmOverloads constructor(
    val ownerId: String,
    val priority: EventPriority = EventPriority.NORMAL,
    val receiveCancelled: Boolean = false,
    val mode: ListenerMode = if (priority == EventPriority.MONITOR) ListenerMode.MONITOR else ListenerMode.MUTATING
) {
    init {
        require(ownerId.isNotBlank()) { "ownerId is required" }
        if (mode == ListenerMode.MUTATING && priority.value > EventPriority.MAX_CUSTOM && ownerId != "pnauth:system") {
            throw IllegalArgumentException("reserved priority")
        }
    }

    companion object {
        @JvmStatic
        fun normal(ownerId: String): ListenerOptions {
            return ListenerOptions(ownerId, EventPriority.NORMAL, false, ListenerMode.MUTATING)
        }

        @JvmStatic
        fun monitor(ownerId: String): ListenerOptions {
            return ListenerOptions(ownerId, EventPriority.MONITOR, true, ListenerMode.MONITOR)
        }
    }
}
